src/qwed_new/core/safe_parser.py

```
@@ -0,0 +1,224 @@
+
+_SAFE_GLOBAL_DICT_TEMPLATE: Dict[str, Any] = {"__builtins__": {}}
+
+
+def _check_ast_depth(expression: str) -> None:
+    """Reject expressions whose AST exceeds max depth (DoS defence)."""
+    try:
+        tree = ast.parse(expression, mode="eval")
+    except SyntaxError:
+        raise SafeParserError(
+            "Expression uses syntax that cannot be validated for safety"
+        )
```

Ruff (0.15.15)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)